src/Common/ExecTrait.php

```diff
@@ -0,0 +1,355 @@
+        return $this;
+    }
+
+    /**
+     * Stops command if it does not output something for a while
+     *
+     * @param int $timeout
+     *
+     * @return $this
+     */
+    public function idleTimeout($timeout)
+    {
+        $this->idleTimeout = $timeout;
+        return $this;
+    }
+
+    /**
+     * Sets the environment variables for the command
+     *
+     * @param array $env
+     *
+     * @return $this
+     */
+    public function env(array $env)
+    {
+        $this->env = $env;
+        return $this;
+    }
+
+    /**
+     * Pass an input to the process. Can be resource created with fopen() or string
+     *
+     * @param resource|string $input
+     *
+     * @return $this
+     */
+    public function setInput($input)
+    {
+        $this->input = $input;
+        return $this;
+    }
+
+    /**
+     * Attach tty to process for interactive input
+     *
+     * @param $interactive bool
+     *
+     * @return $this
+     */
+    public function interactive($interactive)
+    {
+        $this->interactive = $interactive;
+        return $this;
+    }
+
+
+    /**
+     * Is command printing its output to screen
+     *
+     * @return bool
+     */
+    public function getPrinted()
+    {
+        return $this->isPrinted;
+    }
+
+    /**
+     * Changes working directory of command
+     *
+     * @param string $dir
+     *
+     * @return $this
+     */
+    public function dir($dir)
+    {
+        $this->workingDirectory = $dir;
+        return $this;
+    }
+
+    /**
+     * Shortcut for setting isPrinted() and isMetadataPrinted() to false.
+     *
+     * @param bool $arg
+     *
+     * @return $this
+     */
+    public function silent($arg)
+    {
+        if (is_bool($arg)) {
+            $this->isPrinted = !$arg;
+            $this->isMetadataPrinted = !$arg;
+        }
+        return $this;
+    }
+
+    /**
+     * Should command output be printed
+     *
+     * @param bool $arg
+     *
+     * @return $this
+     *
+     * @deprecated
+     */
+    public function printed($arg)
+    {
+        $this->logger->warning("printed() is deprecated. Please use printOutput().");
+        return $this->printOutput($arg);
+    }
+
+    /**
+     * Should command output be printed
+     *
+     * @param bool $arg
+     *
+     * @return $this
+     */
+    public function printOutput($arg)
+    {
+        if (is_bool($arg)) {
+            $this->isPrinted = $arg;
+        }
+        return $this;
+    }
+
+    /**
+     * Should command metadata be printed. I,e., command and timer.
+     *
+     * @param bool $arg
+     *
+     * @return $this
+     */
+    public function printMetadata($arg)
+    {
+        if (is_bool($arg)) {
+            $this->isMetadataPrinted = $arg;
+        }
+        return $this;
+    }
+
+    /**
+     *
+     */
+    public function __destruct()
+    {
+        $this->stop();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function run()
```

Comment:
Can you use ResultData instead?  As you noted, Robo\Result holds the result of a task. `ProcessExecutor` should not be a task, so it should not use a Robo\Result.  `Robo\Task\Base\Exec::run()` should call `ExecTrait::execute()` (or equivalent, which should return a ResultData object. `Exec::run()` should then make a new `Result` object, and populate the fields of the constructor from the ResultData. Maybe a new static constructor in `Result` that creates a new Result from a ResultData would be in order here.

I would rather remove the reference to `$task` from Result entirely, but our backwards compatibility goals (to make 1.x not too different from 0.x) and the way Result & ResultPrinter work precluded this. Maybe in 2.x.

It's important to completely detach ProcessExec from TaskInterface; if ProcessExec is a task, then you might as well just use Exec.